### PR TITLE
Allow to create human readable password (pw token)

### LINKED ISSUE #2864: Allow to define the contents of a generated pw-token
At the moment the pw token has very limited capabilities. You may already enroll a pw token via the api, but you are left with some seed://hexstring password.

I suggest to decouple the password generation (which honors some special contents policy) from the lost token process in
https://github.com/privacyidea/privacyidea/blob/99d10ff1b3cb27af5c9afe3d3e9dd9d1f076fec0/privacyidea/lib/token.py#L1845

I further suggest to return the password like for the registration token with some special field "password".

This change makes the pw token type more useful and may be one of several steps towards application-specific passwords.

### PR DESCRIPTION
**Pull request by cornelinux**
Just like the policies with registration codes
we can now define policies to create human readable
password tokens.

Closes #2864

### CODE REVIEW

**Comment by cornelinux** [comment]
@plettich would you please review and then we could fix the merge conflict.
The problem is, that with every other pull request we will get a merge conflict, since most PRs have changes in lib/policy.py

File: tests/test_api_lib_policy.py [hunk 1]
```diff
@@ -3022,14 +3022,49 @@ def test_36_init_registrationcode_length_contents(self):
...
+        # request, that matches the policy
+        req.all_data = {"user": "cornelius", "realm": "home", "type": "pw"}
+        init_token_length_contents(req)
+        # Check, if the tokenlabel was added
+        self.assertEqual(req.all_data.get("pw.length"), "6")
+        self.assertEqual(req.all_data.get("pw.contents"), "+n")
+        # delete policy
+        delete_policy("pw_length")
+        delete_policy("pw_contents")
+
```
Comment (plettich) [inline tests/test_api_lib_policy.py:3067]: Please also add a test that if the token is neither `pw` nor `registration` we don't have the `lenght` and `content` parameter in `all_data`

File: tests/test_api_validate.py [hunk 2]
```diff
@@ -3101,6 +3116,158 @@ def test_00_registrationtoken_with_validity_period(self):
...
+        with self.app.test_request_context('/token/init',
+                                           method='POST',
+                                           data={'user': 'cornelius',
+                                                 'type': 'registration'},
+                                           headers={'Authorization': self.at}):
+            res = self.app.full_dispatch_request()
+            self.assertTrue(res.status_code == 200, res)
+            data = res.json
+            detail = data.get("detail")
+            regcode = detail.get("registrationcode")
```
Comment (plettich) [inline tests/test_api_validate.py:3135]: we could also check here that the `regcode` has the default length and content

File: tests/test_api_validate.py [hunk 3]
```diff
@@ -3101,6 +3116,158 @@ def test_00_registrationtoken_with_validity_period(self):
...
+                                           method='POST',
+                                           data={'user': 'cornelius',
+                                                 'type': 'registration',
+                                                 'otpkey': "hallo"},
+                                           headers={'Authorization': self.at}):
+            res = self.app.full_dispatch_request()
+            self.assertTrue(res.status_code == 200, res)
+            data = res.json
+            detail = data.get("detail")
+            regcode = detail.get("registrationcode")
```
Comment (plettich) [inline tests/test_api_validate.py:3181]: is the `otpkey` parameter ignored? please check.

File: tests/test_api_validate.py [hunk 4]
```diff
@@ -3101,6 +3116,158 @@ def test_00_registrationtoken_with_validity_period(self):
...
+                                           method='POST',
+                                           data={"user": "cornelius",
+                                                 "pass": regcode}):
+            res = self.app.full_dispatch_request()
+            self.assertTrue(res.status_code == 200, res)
+            data = res.json
+            self.assertEqual("ACCEPT", data.get("result").get("authentication"))
+
+        # The registration code was generated. The passed otpkey was NOT used.
+        with self.app.test_request_context('/validate/check',
```
Comment (plettich) [inline tests/test_api_validate.py:3194]: Ok, we could have just checked the returned `regcode` :-)

File: tests/test_api_validate.py [hunk 5]
```diff
@@ -3101,6 +3116,158 @@ def test_00_registrationtoken_with_validity_period(self):
...
+                                                 'type': 'pw',
+                                                 'genkey': 1,
+                                                 'pin': 'test'},
+                                           headers={'Authorization': self.at}):
+            res = self.app.full_dispatch_request()
+            self.assertEqual(200, res.status_code)
+            data = res.json
+            detail = data.get("detail")
+            serial = detail.get("serial")
+            password = detail.get("password")
```
Comment (plettich) [inline tests/test_api_validate.py:3256]: here we should also check that the password it matches the default length and content

**Review by plettich — CHANGES_REQUESTED** [review]
Mostly test-related changes

File: privacyidea/lib/tokens/passwordtoken.py [hunk 6]
```diff
@@ -132,23 +141,39 @@ def get_class_info(key=None, ret='all'):
...
+            param["otpkey"] = _generate_pin_from_policy(contents, size=int(size))
+        if "otpkey" in param:
+            param["otplen"] = len(param["otpkey"])
         TokenClass.update(self, param)
-        self.set_otplen()
 
     @log_with(log)
     @check_token_locked
-    def set_otplen(self, otplen=0):
+    def set_otplen__(self, otplen=0):
```
[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/lib/tokens/passwordtoken.py:176]: Oh, you are probably right. This is why I named it that funny, I was probably checking if I can get rid of it.

Will do.

File: tests/test_api_lib_policy.py (hunk 1 above)
Comment (cornelinux) [reply tests/test_api_lib_policy.py:3067]: Will check and will do.

File: tests/test_api_validate.py (hunk 2 above)
Comment (cornelinux) [reply tests/test_api_validate.py:3135]: will do.

File: tests/test_api_validate.py (hunk 3 above)
Comment (cornelinux) [reply tests/test_api_validate.py:3181]: Will check.

File: tests/test_api_validate.py (hunk 5 above)
Comment (cornelinux) [reply tests/test_api_validate.py:3256]: will do.

File: tests/test_api_validate.py (hunk 3 above)
Comment (cornelinux) [reply tests/test_api_validate.py:3181]: Done

File: tests/test_api_validate.py (hunk 2 above)
Comment (cornelinux) [reply tests/test_api_validate.py:3135]: Done 
It looks like the default regcode contains special characters. But is so, this was a problem, that already existed earlier and we should recheck this and open another issue for that.

File: tests/test_api_lib_policy.py (hunk 1 above)
Comment (cornelinux) [reply tests/test_api_lib_policy.py:3067]: Done

File: privacyidea/lib/tokens/passwordtoken.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/lib/tokens/passwordtoken.py:176]: Done

File: tests/test_api_validate.py (hunk 4 above)
Comment (cornelinux) [reply tests/test_api_validate.py:3203]: There is nothing to do here, right?

File: tests/test_api_validate.py (hunk 2 above)
Comment (plettich) [reply tests/test_api_validate.py:3137]: But according to
https://github.com/privacyidea/privacyidea/blob/4dff583f3e099dd1479b2454115a83c6700e9f93/privacyidea/lib/tokens/registrationtoken.py#L41
the contents should have been "cn". Where did You encounter special characters?

Comment (plettich) [reply tests/test_api_validate.py:3137]: Maybe this is the reason why the `APITokenTestCase.test_15_registration_code()` fails sometimes.

File: tests/test_api_validate.py [hunk 7]
```diff
@@ -3101,6 +3118,170 @@ def test_00_registrationtoken_with_validity_period(self):
...
+                                                 'type': 'registration'},
+                                           headers={'Authorization': self.at}):
+            res = self.app.full_dispatch_request()
+            self.assertTrue(res.status_code == 200, res)
+            data = res.json
+            detail = data.get("detail")
+            regcode = detail.get("registrationcode")
+            self.assertEqual(DEFAULT_LENGTH_REG, len(regcode))
+            # Check if a number is contained
+            self.assertRegex(regcode, "[0-9]+")
```
Comment (plettich) [inline tests/test_api_validate.py:3140]: Please use `assertRegexpMatches()` here, even though it is deprecated, we can check with Python 2 as well.

File: tests/test_api_validate.py [hunk 8]
```diff
@@ -3101,6 +3118,170 @@ def test_00_registrationtoken_with_validity_period(self):
...
+                                           headers={'Authorization': self.at}):
+            res = self.app.full_dispatch_request()
+            self.assertEqual(200, res.status_code)
+            data = res.json
+            detail = data.get("detail")
+            serial = detail.get("serial")
+            password = detail.get("password")
+            self.assertEqual(DEFAULT_LENGTH_PW, len(password))
+            # Check if a number is contained
+            self.assertRegex(password, "[0-9]+")
```
Comment (plettich) [inline tests/test_api_validate.py:3268]: See above

**Review by plettich — CHANGES_REQUESTED** [review]

**Review by plettich — APPROVED** [review]
looks good

**Comment by plettich** [comment]
Closed in favor of PR #3023

**Pull request closed without merging**